File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        start = 0
        end = len(nums)-1
        while (start<end):
            mid = start + (end-start)//2
            #1st branch
            if (nums[0]<=nums[mid]):
                start = mid +1
            else:
                end = mid
        #now start=end=mid is teh pivot value
        if (target >= nums[start] and target <= nums[len(nums)-1]) :
            s = start
            e = len(nums)-1
        else:
            s = 0
            e = start-1
        while (s<=e):
            m = s + (e-s)//2
            if (nums[m]==target):
                return m
            elif (nums[m]<target):
                s = m+1
            else:
                e = m-1
        return -1

        
        
    

                
            


        

File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int start = 0;
        int end = nums.size() - 1;

        // Step 1: Find pivot (index of smallest element)
        while (start < end) {
            int mid = start + (end - start) / 2;

            if (nums[0] <= nums[mid]) {
                start = mid + 1;
            } else {
                end = mid;
            }
        }

        int pivot = start;

        // Step 2: Choose which half to search
        int s, e;

        if (target >= nums[pivot] && target <= nums[nums.size() - 1]) {
            s = pivot;
            e = nums.size() - 1;
        } else {
            s = 0;
            e = pivot - 1;
        }

        // Step 3: Normal binary search
        while (s <= e) {
            int m = s + (e - s) / 2;

            if (nums[m] == target) {
                return m;
            }
            if (target < nums[m]) {
                e = m - 1;
            } else {
                s = m + 1;
            }
        }

        return -1;
    }
};
